### ISSUE TITLE
Overlay network

### ISSUE THREAD

**Issue by cicdteam** [neondatabase/autoscaling#104]
## Access to test EKS cluster with NeonVM Overlay Network

```console
aws --profile dev sso login
aws --profile dev  eks --region us-east-2 update-kubeconfig --name dev-us-east-2-andrey --alias dev-us-east-2-andrey
```

### vm-1

```yaml
apiVersion: vm.neon.tech/v1
kind: VirtualMachine
metadata:
  name: vm-1
spec:
  terminationGracePeriodSeconds: 1
  nodeSelector:
    node: neon1
  guest:
    cpus:
      min: 1
    memorySlotSize: 1Gi
    memorySlots:
      min: 1
    rootDisk:
      image: andreyneon/vm-nginx:alpine
    ports:
      - name: http
        port: 80
  extraNetwork:
      enable: true
```

### vm-2

```yaml
apiVersion: vm.neon.tech/v1
kind: VirtualMachine
metadata:
  name: vm-2
spec:
  terminationGracePeriodSeconds: 1
  nodeSelector:
    node: neon2
  guest:
    cpus:
      min: 1
    memorySlotSize: 1Gi
    memorySlots:
      min: 1
    rootDisk:
      image: andreyneon/vm-nginx:alpine
    ports:
      - name: http
        port: 80
  extraNetwork:
      enable: true
```

### pod (just pod)

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod
spec:
  terminationGracePeriodSeconds: 1
  nodeSelector:
    node: neon3
  containers:
  - name: pod
    image: nginx:alpine
    imagePullPolicy: IfNotPresent
    ports:
      - name: http
        containerPort: 80
    securityContext:
      capabilities:
        add:
        - NET_RAW
        - NET_ADMIN
```

## Access to EKS nodes

node where `vm-1` running

```console
aws --profile dev --region us-east-2 ssm start-session --target i-04c441ec6d78200dc 
```

node where `vm-2` running

```console
aws --profile dev --region us-east-2 ssm start-session --target i-0d532401c1ede5ac5
```

node where test `pod` running

```console
aws --profile dev --region us-east-2 ssm start-session --target i-053b0d713ebec6aed
```

## Network structure

See attached picture

- each node has VXLAN configured:
  - linux bridge interface `neon-br0` with IP from Overlay CIDR `10.100.0.0/20`, linked to `neon-vxlan0` interface
  - `neon-vxlan0` interface as part of overlay VXLAN network
- each runner pod (with VM inside) has several interfaces:
  - `eth0` - belongs to standard Pod network (IP from AWS VPC CIDR)
  - `net1` - belongs to Overlay VXLAN Network (linked to `neon-br0` interface on host), IP from Overlay CIDR `10.100.0.0/20`
  - `br-def` and `tap-def` - linux bridge and tuntap interface used in QEMU for `eth0` inside VM
  - `br-overlay` and `tap-overlay` - linux bridge and tuntap interface used in QEMU for `eth1` inside VM
- each VM has two interfaces:
  - `eth0` - communicate with standard Pod network
  - `eth1` - communicate with Overlay network (and has IP from Overlay CIDR)

### Pings from `vm-1`

```console
% k get neonvm
NAME   CPUS   MEMORY   POD          EXTRAIP      STATUS    AGE
vm-1   1      1Gi      vm-1-vbq8r   10.100.0.4   Running   126m
vm-2   1      1Gi      vm-2-47dxp   10.100.0.6   Running   126m

% k exec -it vm-1-vbq8r -- screen /dev/pts/0

neonvm:~# # pinging vm-2 overlay IP
neonvm:~#
neonvm:~# ping 10.100.0.6 -c 4
PING 10.100.0.6 (10.100.0.6): 56 data bytes
64 bytes from 10.100.0.6: seq=0 ttl=64 time=13.030 ms
64 bytes from 10.100.0.6: seq=1 ttl=64 time=1.651 ms
64 bytes from 10.100.0.6: seq=2 ttl=64 time=1.959 ms
64 bytes from 10.100.0.6: seq=3 ttl=64 time=1.804 ms

--- 10.100.0.6 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 1.651/4.611/13.030 ms
neonvm:~#

neonvm:~# # pinging overlay IP of some k8s node
neonvm:~#
neonvm:~# ping 10.100.0.1 -c 4
PING 10.100.0.1 (10.100.0.1): 56 data bytes
64 bytes from 10.100.0.1: seq=0 ttl=255 time=11.614 ms
64 bytes from 10.100.0.1: seq=1 ttl=255 time=1.216 ms
64 bytes from 10.100.0.1: seq=2 ttl=255 time=1.245 ms
64 bytes from 10.100.0.1: seq=3 ttl=255 time=1.239 ms

--- 10.100.0.1 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 1.216/3.828/11.614 ms
neonvm:~#

neonvm:~# # pinging outside resource
neonvm:~#
neonvm:~# ping 8.8.8.8 -c 4
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=103 time=13.653 ms
64 bytes from 8.8.8.8: seq=1 ttl=103 time=12.483 ms
64 bytes from 8.8.8.8: seq=2 ttl=103 time=12.420 ms
64 bytes from 8.8.8.8: seq=3 ttl=103 time=12.497 ms

--- 8.8.8.8 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 12.420/12.763/13.653 ms
neonvm:~#

neonvm:~# # pinging test pod IP
neonvm:~#
neonvm:~# ping 10.99.39.37 -c 4
PING 10.99.39.37 (10.99.39.37): 56 data bytes
64 bytes from 10.99.39.37: seq=0 ttl=252 time=4.934 ms
64 bytes from 10.99.39.37: seq=1 ttl=252 time=1.583 ms
64 bytes from 10.99.39.37: seq=2 ttl=252 time=1.584 ms
64 bytes from 10.99.39.37: seq=3 ttl=252 time=1.647 ms

--- 10.99.39.37 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 1.583/2.437/4.934 ms
```

### Pings from `vm-2`

...all the same

### Pings from test pod

```console
/ # # pinging outside resource
/ #
/ # ping 8.8.8.8 -c 4
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: seq=0 ttl=104 time=12.942 ms
64 bytes from 8.8.8.8: seq=1 ttl=104 time=12.670 ms
64 bytes from 8.8.8.8: seq=2 ttl=104 time=12.686 ms
64 bytes from 8.8.8.8: seq=3 ttl=104 time=12.707 ms

--- 8.8.8.8 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 12.670/12.751/12.942 ms
/ #


/ # # pinging VM-1 pod network IP
/ #
/ # ping 10.99.12.77 -c 4
PING 10.99.12.77 (10.99.12.77): 56 data bytes
64 bytes from 10.99.12.77: seq=0 ttl=253 time=1.119 ms
64 bytes from 10.99.12.77: seq=1 ttl=253 time=1.077 ms
64 bytes from 10.99.12.77: seq=2 ttl=253 time=1.102 ms
64 bytes from 10.99.12.77: seq=3 ttl=253 time=1.119 ms

--- 10.99.12.77 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 1.077/1.104/1.119 ms
/ #

/ # ping 10.100.0.4 -c 4 -W 1
PING 10.100.0.4 (10.100.0.4): 56 data bytes

--- 10.100.0.4 ping statistics ---
4 packets transmitted, 0 packets received, 100% packet loss
/ #
```

### Firstly try to ping from `vm-2` to `vm-1` (via overlay) and watch tcpdump on `vm-1`

ping from `vm-2` to `vm-1`

```console
neonvm:~#
neonvm:~# ping 10.100.0.4 -c 4
PING 10.100.0.4 (10.100.0.4): 56 data bytes
64 bytes from 10.100.0.4: seq=0 ttl=64 time=8.184 ms
64 bytes from 10.100.0.4: seq=1 ttl=64 time=1.755 ms
64 bytes from 10.100.0.4: seq=2 ttl=64 time=1.574 ms
64 bytes from 10.100.0.4: seq=3 ttl=64 time=1.695 ms

--- 10.100.0.4 ping statistics ---
4 packets transmitted, 4 packets received, 0% packet loss
round-trip min/avg/max = 1.574/3.302/8.184 ms
neonvm:~#
```

tcpdump on `vm-1`

```console
neonvm:~# tcpdump -i eth1 -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth1, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:00:49.926544 IP 10.100.0.6 > 10.100.0.4: ICMP echo request, id 18816, seq 0, length 64
20:00:49.926832 IP 10.100.0.4 > 10.100.0.6: ICMP echo reply, id 18816, seq 0, length 64
20:00:50.928202 IP 10.100.0.6 > 10.100.0.4: ICMP echo request, id 18816, seq 1, length 64
20:00:50.928360 IP 10.100.0.4 > 10.100.0.6: ICMP echo reply, id 18816, seq 1, length 64
20:00:51.928583 IP 10.100.0.6 > 10.100.0.4: ICMP echo request, id 18816, seq 2, length 64
20:00:51.928745 IP 10.100.0.4 > 10.100.0.6: ICMP echo reply, id 18816, seq 2, length 64
20:00:52.929105 IP 10.100.0.6 > 10.100.0.4: ICMP echo request, id 18816, seq 3, length 64
20:00:52.929266 IP 10.100.0.4 > 10.100.0.6: ICMP echo reply, id 18816, seq 3, length 64

8 packets captured
8 packets received by filter
0 packets dropped by kernel
neonvm:~#
```

### Now ping `vm-1` from `test pod` and watch tcpdumps everywhere

ping from `test pod` to `vm-1`

```console
/ # ping 10.100.0.4 -c 4 -W 1
PING 10.100.0.4 (10.100.0.4): 56 data bytes

--- 10.100.0.4 ping statistics ---
4 packets transmitted, 0 packets received, 100% packet loss
/ #
```

tcpdump in `vm-1`, `eth1` interface

```console
neonvm:~# tcpdump -i eth1 -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth1, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:06:40.571929 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 62, seq 0, length 64
20:06:41.571998 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 62, seq 1, length 64
20:06:42.572059 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 62, seq 2, length 64
20:06:43.572122 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 62, seq 3, length 64
```

tcpdump in `vm-1`, but `eth0` interface (as outgoing traffic should go to the Pod network)

```console
neonvm:~# tcpdump -i eth0 -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:08:11.481229 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 63, seq 0, length 64
20:08:12.471460 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 63, seq 1, length 64
20:08:13.471544 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 63, seq 2, length 64
20:08:14.471612 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 63, seq 3, length 64
```

So... we see icmp-requests arrive to `vm-1` via `eth1` and icmp-reply leave `vm-1` via `eth0`, seems all correct...

But why replies do not reach `test pod` ? Try to see tcpdumps in `vm-1` runner pod interfaces

tcpdump in `vm-1`'s runner pod, interface `tap-overlay` (it used as net device in QEMU for `eth1` in VM)

```console
/ # tcpdump -i tap-overlay -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on tap-overlay, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:13:58.069967 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 65, seq 0, length 64
20:13:59.068572 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 65, seq 1, length 64
20:14:00.068640 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 65, seq 2, length 64
20:14:01.068716 IP 10.99.39.37 > 10.100.0.4: ICMP echo request, id 65, seq 3, length 64
```

so.. we see incoming icmp requests go via `tap-overlay` to the VM as expected, but where is replies?

tcpdump in `vm-1`'s runner pod, interface `tap-def` (it used as net device in QEMU for `eth0` in VM)

```console
/ # tcpdump -i tap-def -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on tap-def, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:16:16.816267 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 66, seq 0, length 64
20:16:17.815074 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 66, seq 1, length 64
20:16:18.815202 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 66, seq 2, length 64
20:16:19.815226 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 66, seq 3, length 64
```

so, we see replies leave VM and QEMU... let check `br-def` then...

```console
/ # tcpdump -i br-def -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on br-def, link-type EN10MB (Ethernet), snapshot length 262144 bytes
20:17:14.308064 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 67, seq 0, length 64
20:17:15.301010 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 67, seq 1, length 64
20:17:16.301083 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 67, seq 2, length 64
20:17:17.301159 IP 10.100.0.4 > 10.99.39.37: ICMP echo reply, id 67, seq 3, length 64
```

and we see replies go to the bridge... so, now they should leave pod (respecting route table they shoud leave via default route)

```console
/ # tcpdump -i eth0 -n icmp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
^C
0 packets captured
0 packets received by filter
0 packets dropped by kernel
/ #
```

but... no any packets on `eth0` in runner pod... :(

![NeonVM-OverlayNet](https://user-images.githubusercontent.com/14281009/220461351-6172251a-67fa-49d6-a09f-5b36eebb969e.jpg)

**Comment by sharnoff** [comment]
@cicdteam this was implemented/fixed by #256, right?

**Comment by sharnoff** [comment]
This was marked as done on 2023-07-12, but not actually closed.
